### PR TITLE
[Azure AI HealthInsights - RadiologyInsigts] Java SDK version V2

### PR DESCRIPTION
**Pull request by asaflevi-ms**
# Description

Radiology Insights is a new model/skill that is going for public in V2 of HealthInsights public service .

Typespec PR: https://github.com/Azure/azure-rest-api-specs/pull/27223

# All SDK Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/healthinsights/pom.xml [hunk 1]
```diff
@@ -10,5 +10,6 @@
   <modules>
       <module>azure-health-insights-cancerprofiling</module>
       <module>azure-health-insights-clinicalmatching</module>
+      <module>azure-health-insights-radiologyinsights</module>
```
Comment (weidongxu-microsoft) [inline sdk/healthinsights/pom.xml:13]: It also need to be added to ci.yml (should be at same folder).

Comment (tomsft) [reply sdk/healthinsights/pom.xml:13]: Added it to ci.yml with latest commit to the branch.

File: sdk/healthinsights/azure-health-insights-radiologyinsights/src/samples/java/com/azure/health/insights/radiologyinsights/SampleCriticalResultInferenceAsync.java [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+                mono = completedResult.getFinalResult();
+            } else {
+                System.out.println(
+                        "Poll Request Failed with message: " + completedResult.getValue().getError().getMessage());
+            }
+            latch.countDown();
+        });
+
+        latch.await();
+        displayCriticalResults(mono.block());
```
Comment (weidongxu-microsoft) [inline sdk/healthinsights/azure-health-insights-radiologyinsights/src/samples/java/com/azure/health/insights/radiologyinsights/SampleCriticalResultInferenceAsync.java:111]: The reactor code may not be best practise.

candidate
```java
        Mono<RadiologyInsightsInferenceResult> finalResult = asyncPoller.last().flatMap(pollResult -> {
            if (pollResult.getStatus() == LongRunningOperationStatus.SUCCESSFULLY_COMPLETED) {
                return pollResult.getFinalResult();
            } else {
                System.out.println(
                        "Poll Request Failed with message: " + pollResult.getValue().getError().getMessage());
                return Mono.empty();    // Mono.error ?
            }
        });

        finalResult.doFinally(() -> {
            latch.countDown();
        }).subscribe(result -> {
            displayCriticalResults(result);
        });

        latch.countDown();
```

File: sdk/healthinsights/azure-health-insights-radiologyinsights/README.md [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    <groupId>com.azure</groupId>
+    <artifactId>azure-health-insights-radiologyinsights</artifactId>
+    <version>1.0.0-beta.1</version>
+</dependency>
+```
+[//]: # ({x-version-update-end})
+
+### Authenticate the client
+
+In order to interact with the Health Insights Clinical Matching service, you'll need to create an instance of the RadiologyInsightsClient class.  You will need an **endpoint** and an **API key** to instantiate a client object.
```
Comment (weidongxu-microsoft) [inline sdk/healthinsights/azure-health-insights-radiologyinsights/README.md:32]: "Health Insights Clinical Matching service" <-- please proofread the doc.

File: sdk/healthinsights/azure-health-insights-radiologyinsights/tsp-location.yaml [hunk 4]
```diff
@@ -0,0 +1,7 @@
+directory: specification/ai/HealthInsights/HealthInsights.RadiologyInsights
+additionalDirectories:
+  - specification/ai/HealthInsights/HealthInsights.Common
+  - specification/ai/HealthInsights/HealthInsights.OpenAPI
+commit: 3efb341e0472b4b1d85f9c015666a715d9f5fa5a
```
[the remark this answers is not in the document]
Comment (tomsft) [reply sdk/healthinsights/azure-health-insights-radiologyinsights/tsp-location.yaml:5]: It's on the main branch:
$ git branch --contains 3efb341e0472b4b1d85f9c015666a715d9f5fa5a
* main

File: sdk/healthinsights/azure-health-insights-radiologyinsights/src/samples/java/com/azure/health/insights/radiologyinsights/SampleCriticalResultInferenceAsync.java [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+            + "\r\nThe uterus is unremarkable given the transabdominal technique with "
+            + "\r\nendometrial echo complex within physiologic normal limits. The "
+            + "\r\novaries are symmetric in size, measuring 2.5 x 1.2 x 3.0 cm and the "
+            + "\r\nleft measuring 2.8 x 1.5 x 1.9 cm.\n \r\nOn duplex imaging, Doppler signal is symmetric."
+            + "\r\n "
+            + "\r\nIMPRESSION:   "
+            + "\r\n1. Normal pelvic sonography. Findings of testicular torsion."
+            + "\r\n\nA new US pelvis within the next 6 months is recommended."
+            + "\n\nThese results have been discussed with Dr. Jones at 3 PM on November 5 2020.\n "
+            + "\r\n";
```
[the remark this answers is not in the document]
Comment (tomsft) [reply sdk/healthinsights/azure-health-insights-radiologyinsights/src/samples/java/com/azure/health/insights/radiologyinsights/SampleCriticalResultInferenceAsync.java:80]: This is a mock-up example, doesn't contain PHI.

File: sdk/healthinsights/azure-health-insights-radiologyinsights/src/samples/java/com/azure/health/insights/radiologyinsights/SampleCriticalResultInferenceAsync.java (hunk 2 above)
Comment (tomsft) [reply sdk/healthinsights/azure-health-insights-radiologyinsights/src/samples/java/com/azure/health/insights/radiologyinsights/SampleCriticalResultInferenceAsync.java]: refactored to reflect suggested code.

File: sdk/healthinsights/azure-health-insights-radiologyinsights/src/samples/java/com/azure/health/insights/radiologyinsights/SampleCriticalResultInferenceAsync.java [hunk 6]
```diff
@@ -0,0 +1,287 @@
...
+                        "Poll Request Failed with message: " + completedResult.getValue().getError().getMessage());
+            }
+            latch.countDown();
+        });
+
+        latch.await();
+        displayCriticalResults(mono.block());
+    }
+
+    private static Mono<RadiologyInsightsInferenceResult> mono = null;
```
[the remark this answers is not in the document]
Comment (tomsft) [reply sdk/healthinsights/azure-health-insights-radiologyinsights/src/samples/java/com/azure/health/insights/radiologyinsights/SampleCriticalResultInferenceAsync.java]: with refactor got rid of the class variables and moved them to function.

File: sdk/healthinsights/azure-health-insights-radiologyinsights/README.md (hunk 3 above)
Comment (tomsft) [reply sdk/healthinsights/azure-health-insights-radiologyinsights/README.md]: Corrected the readme file and made some more changes with references to the services published documentation.

File: sdk/healthinsights/azure-health-insights-radiologyinsights/README.md [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+```bash
+az cognitiveservices account keys list --resource-group <your-resource-group-name> --name <your-resource-name>
+```
+
+#### Create a RadiologyInsightsClient with an API Key Credential
+
+Once you have the value for the API key, you can pass it as a string into an instance of **AzureKeyCredential**. Use the key as the credential parameter
+to authenticate the client:
+
+```java
```
[the remark this answers is not in the document]
Comment (tomsft) [reply sdk/healthinsights/azure-health-insights-radiologyinsights/README.md]: This is fixed in the last commit, added Code Snippet markers to the Sample java code, added the code snippet plugin to the pom file, and generated the java code snippets in the README.md through a maven build.

File: sdk/healthinsights/azure-health-insights-radiologyinsights/README.md [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+## Examples
+
+### Create a RadiologyInsights request and get the result using an asynchronous client
+
+Infer radiology insights from a patient's radiology report.
+
+```java com.azure.health.insights.radiologyinsights.inferradiologyinsights
+PollerFlux<PollOperationDetails, RadiologyInsightsInferenceResult> asyncPoller = radiologyInsightsAsyncClient
+        .beginInferRadiologyInsights(createRadiologyInsightsRequest());
+```
```
Comment (weidongxu-microsoft) [inline sdk/healthinsights/azure-health-insights-radiologyinsights/README.md:72]: nit, usually SDK would first give example code on sync client

(user not familiar with Reactor likely found it difficult to use async client)

File: sdk/healthinsights/azure-health-insights-radiologyinsights/README.md [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+Once you have the value for the API key, you can pass it as a string into an instance of **AzureKeyCredential**. Use the key as the credential parameter
+to authenticate the client:
+
+```java com.azure.health.insights.radiologyinsights.buildasyncclient
+String endpoint = Configuration.getGlobalConfiguration().get("AZURE_HEALTH_INSIGHTS_ENDPOINT");
+String apiKey = Configuration.getGlobalConfiguration().get("AZURE_HEALTH_INSIGHTS_API_KEY");
+
+RadiologyInsightsAsyncClient radiologyInsightsAsyncClient = new RadiologyInsightsClientBuilder()
+        .endpoint(endpoint).serviceVersion(RadiologyInsightsServiceVersion.getLatest())
+        .credential(new AzureKeyCredential(apiKey)).buildAsyncClient();
```
Comment (weidongxu-microsoft) [inline sdk/healthinsights/azure-health-insights-radiologyinsights/README.md:56]: Same here, better to let user build `RadiologyInsightsClient`.

File: sdk/healthinsights/azure-health-insights-radiologyinsights/README.md (hunk 8 above)
Comment (weidongxu-microsoft) [reply sdk/healthinsights/azure-health-insights-radiologyinsights/README.md:72]: And as an example, the snippet would be better to include the code in `createRadiologyInsightsRequest`.

Else, reader probably have no idea how to create the `RadiologyInsightsData` object.

File: sdk/healthinsights/azure-health-insights-radiologyinsights/pom.xml [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+  <developers>
+    <developer>
+      <id>microsoft</id>
+      <name>Microsoft</name>
+    </developer>
+  </developers>
+  <properties>
+    <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+    <jacoco.min.linecoverage>0.1</jacoco.min.linecoverage>
+    <jacoco.min.branchcoverage>0.1</jacoco.min.branchcoverage>
```
Comment (weidongxu-microsoft) [inline sdk/healthinsights/azure-health-insights-radiologyinsights/pom.xml:47]: nit, current coverage seems > 0.2
https://dev.azure.com/azure-sdk/public/_build/results?buildId=3464063&view=codecoverage-tab

**Review by weidongxu-microsoft — COMMENTED** [review]
Mostly LGTM

File: sdk/healthinsights/azure-health-insights-radiologyinsights/README.md (hunk 8 above)
Comment (tomsft) [reply sdk/healthinsights/azure-health-insights-radiologyinsights/README.md]: Added both the sync and async example code to the readme file with precedence for sync call. The full request creation has also been included in the sample with last commit.

File: sdk/healthinsights/azure-health-insights-radiologyinsights/README.md (hunk 9 above)
Comment (tomsft) [reply sdk/healthinsights/azure-health-insights-radiologyinsights/README.md:69]: Added example code to the README file for the synchronous client in last commit.

File: sdk/healthinsights/azure-health-insights-radiologyinsights/README.md [hunk 11]
```diff
@@ -0,0 +1,283 @@
...
+    patientDetails.setBirthDate(dateTime.toLocalDate());
+    
+    patientRecord.setInfo(patientDetails);
+
+    Encounter encounter = new Encounter("encounterid1");
+
+    TimePeriod period = new TimePeriod();
+    DateTimeFormatter formatter2 = DateTimeFormatter.ofPattern("yyyy-M-d'T'HH:mm:ssXXX");
+
+    OffsetDateTime startTime = OffsetDateTime.parse("2021-8-28T00:00:00" + "+00:00", formatter2);
```
Comment (weidongxu-microsoft) [inline sdk/healthinsights/azure-health-insights-radiologyinsights/README.md:130]: nit, I assume `OffsetDateTime.parse("2021-08-28T00:00:00Z")` maybe simpler?

File: sdk/healthinsights/azure-health-insights-radiologyinsights/README.md [hunk 12]
```diff
@@ -0,0 +1,283 @@
...
+    PatientRecord patientRecord = new PatientRecord("Sharona");
+
+    PatientDetails patientDetails = new PatientDetails();
+    patientDetails.setSex(PatientSex.FEMALE);
+    // Define a formatter that matches the input pattern
+    DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ssXXX");
+
+    // Parse the string to LocalDateTime
+    LocalDateTime dateTime = LocalDateTime.parse("1959-11-11T19:00:00+00:00", formatter);
+    patientDetails.setBirthDate(dateTime.toLocalDate());
```
Comment (weidongxu-microsoft) [inline sdk/healthinsights/azure-health-insights-radiologyinsights/README.md:121]: nit, `LocalDate.of(1959, 11, 11)` maybe simpler

**Comment by weidongxu-microsoft** [comment]
From teams chat, please try add a src/samples/README.md

A simple one would do, just link to file, and what it does.
(note there could be CI report on link because it's not in main; if so, you can add link after the PR merged)

Example:
- https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/contentsafety/azure-ai-contentsafety/src/samples/README.md
- https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/vision/azure-ai-vision-imageanalysis/src/samples/README.md

**Review by weidongxu-microsoft — APPROVED** [review]

**Pull request merged**